lib/backends/webgl/inference-handler.ts

```diff
@@ -5,105 +5,151 @@ import {InferenceHandler} from '../../backend';
-import {TextureData, TextureLayout} from './texture-data';
 import {Encoder} from './texture-data-encoder';
 import {TextureHelper} from './texture-helper';
 import {WidthHeightPrefs} from './texture-layout-strategy';
+import {TextureData, TextureLayout, WebGLOperator} from './types';
 import {getPackedShape} from './utils';
 
-/**
- * GlInferencContext is reponsible for mapping from Tensors to TextureData
- * and back
- * Throughout WebGL backend operations TextureData is used as the data carrier
- */
 export class WebGLInferenceHandler implements InferenceHandler {
   textureHelper: TextureHelper;
   programManager: ProgramManager;
-  private tensorToTexture: Map<Tensor, TextureData>;
-  private textureToTensor: Map<TextureData, Tensor>;
-  constructor(public backend: WebGLBackend, public session: WebGLSessionHandler) {
+  private textureDataCache: Map<Tensor.Id, TextureData>;
+  constructor(public session: WebGLSessionHandler) {
     this.textureHelper = session.textureHelper;
     this.programManager = session.programManager;
-    this.tensorToTexture = new Map();
-    this.textureToTensor = new Map();
+    this.textureDataCache = new Map();
   }
-  protected lookupTextureData(tensor: Tensor): TextureData|undefined {
-    const isInitializer = this.session.isInitializer(tensor);
-    Logger.verbose('InferenceHandler', `tensor was an initializer; returning TextureData from session cache`);
-    return isInitializer ? this.session.getTextureData(tensor) : this.tensorToTexture.get(tensor);
+
+  run(op: WebGLOperator, inputs: Tensor[]): Tensor[] {
+    let artifact = this.programManager.getArtifact(op);
+    if (!artifact) {
+      const programInfo = op.createProgramInfo(this, inputs);
+      artifact = this.programManager.build(programInfo);
+      this.programManager.setArtifact(op, artifact);
+    }
+    const runData = op.createRunData(this, artifact.programInfo, inputs);
+    this.programManager.run(artifact, runData);
+    return [runData.outputTextureData.tensor];
   }
-  getOrCreate(tensor: Tensor, layout?: TextureLayout): TextureData {
-    let td = this.lookupTextureData(tensor);
+
+  /**
+   * Create a TextureData object from a tensor.
+   * Usage = Encoder.Usage.UploadOnly.
+   * If a related texture data is found in cache, returns it;
+   * Otherwise:
+   *   Creates a new texture layout if not provided;
+   *   Creates WebGLTexture with the layout;
+   *   Upload tensor data to the texture;
+   *   Creates a texture data object associated with the given tensor.
+   * @param tensor the tensor with data to upload
+   */
+  getOrCreateTextureData(tensor: Tensor, layout?: TextureLayout) {
+    let td = this.getTextureData(tensor.dataId);
     if (!td) {
       Logger.verbose('InferenceHandler', `Creating new TextureData for dims: [${tensor.dims}]`);
       if (!layout) {
-        layout = this.createBasicTextureLayout(tensor.dims.slice());
+        layout = this.createTextureLayoutFromShape(tensor.dims.slice());
```

Comment:
Do we need this `slice()` ?

Reply:
close those duplicated comments related to `slice()`